core/src/main/java/org/elasticsearch/index/query/support/InnerHitsQueryParserHelper.java

```diff
@@ -130,4 +132,20 @@ public static void parseCommonInnerHitOptions(XContentParser parser, XContentPar
             }
         }
     }
+
+    private static void parseHighlight(XContentParser parser, SearchContext context) {
+        QueryParseContext parseContext = context.getQueryShardContext( ).parseContext();
+        XContentParser old = parseContext.parser();
```

Comment:
Yes, I traced this back to `InnerHitsQueryParserHelper#parse` and `InnerHitsParseElement#parseSubSearchContext` where new SubSearchContexts are created. The context they wrap might hold the same parser than the one that is passed down, but it's hard to trace this, so in order ot make sure when parsing the Hightlighter we use the parser that is passed in I'm doing the reset/restore dance here.